File: mobile/src/hiddenApiStubs/java/android/net/wifi/WifiClient.java
package android.net.wifi;

import android.net.MacAddress;
import android.os.Parcelable;
import androidx.annotation.RequiresApi;

@RequiresApi(30)
public abstract class WifiClient implements Parcelable {
    @RequiresApi(31)
    public abstract String getApInstanceIdentifier();

    public abstract int getDisconnectReason();

    public abstract MacAddress getMacAddress();
}

File: mobile/src/main/java/be/mygod/vpnhotspot/net/wifi/WifiClient.kt
package be.mygod.vpnhotspot.net.wifi

import android.net.wifi.WifiClient
import android.os.Build
import androidx.annotation.RequiresApi
import be.mygod.vpnhotspot.util.UnblockCentral
import timber.log.Timber

@get:RequiresApi(31)
val WifiClient.apInstanceIdentifierOrNull: String? get() = try {
    UnblockCentral.getApInstanceIdentifier(this)
} catch (e: NoSuchMethodError) {
    if (Build.VERSION.SDK_INT >= 31) Timber.w(e)
    null
}
